packages/ai-jsx/src/batteries/constrained-output.tsx

```diff
@@ -5,169 +5,310 @@
-          {typeName.toLocaleLowerCase()} ``` code blocks.
+          You must reformat the string to be a valid {props.typeName} object, but you must keep the same data.
         </UserMessage>
-      </ChatCompletion>
+      </OneShotObjectCompletion>
     );
 
-    output = await render(completionRetry);
-    valiationResults = validator(output);
-    if (valiationResults.success) {
+    try {
+      output = yield* render(completionRetry);
       return output;
+    } catch (e: any) {
+      validationError = e.message;
     }
 
     logger.debug(
-      { atempt: retryIndex + 1, expectedFormat: typeName, output },
-      `Output did not validate to ${typeName}.`
+      { atempt: retryIndex + 1, expectedFormat: props.typeName, output },
+      `Output did not validate to ${props.typeName}.`
     );
   }
 
   throw new AIJSXError(
-    `The model did not produce a valid ${typeName} object, even after ${retries} attempts.`,
+    `The model did not produce a valid ${props.typeName} object, even after ${retries} attempts.`,
     ErrorCode.ModelOutputDidNotMatchConstraint,
     'runtime',
     {
-      typeName,
+      typeName: props.typeName,
       retries,
       output,
     }
   );
 }
 
-function isJsonString(str: string): ValidationResult {
-  try {
-    JSON.parse(str);
-    return { success: true, error: '' };
-  } catch (e: any) {
-    return { success: false, error: e.message };
-  }
-}
+/**
+ * A {@link ChatCompletion} component that constrains the output to be a valid JSON string.
+ * It (ab)uses OpenAI function calls to generate the JSON string.
+ *
+ * @returns A string that is a valid JSON or throws an error.
+ */
+export async function JsonChatCompletionFunctionCall(
+  { schema, children, ...props }: ModelPropsWithChildren & { schema: z.Schema },
+  { render }: AI.ComponentContext
+) {
+  const childrenWithCompletion = (
+    <ChatCompletion
+      {...props}
+      functionDefinitions={{
+        print: {
+          description: 'Prints the response in a human readable format.',
+          parameters: schema,
+        },
+      }}
+    >
+      {children}
+      <SystemMessage>
+        Your response must use the `print` function that is provided. No other explanation needed.
+      </SystemMessage>
+    </ChatCompletion>
+  );
 
-function isYamlString(str: string): ValidationResult {
-  try {
-    yaml.load(str);
-    return { success: true, error: '' } as ValidationResult;
-  } catch (e: any) {
-    return { success: false, error: e.message } as ValidationResult;
+  const output = await render(childrenWithCompletion, { stop: (e) => e.tag == FunctionCall });
+
+  // TODO: can we make this work with progressive loading? This will require changes to core/completion.
+
+  const lastYield = output[output.length - 1];
+  if (AI.isElement(lastYield) && lastYield.tag == FunctionCall) {
+    console.log('args', lastYield.props.args);
+    // TODO: is validation necessary? If yes, how to handle errors?
```

Comment:
maybe similar to how you have `<ObjectCompletionWithRetry>` and `<OneShotObjectCompletion>`, we can have the same separation here and we pass the parsing error as a `<FunctionResponse>` back to the model.